Feature/ratt

wouterbeek opened the pull request:
This pull request includes the following changes:

- Small generic README file to explain what this repo is about.
- Generic pages for clean code, editors, environment variables, and API Tokens.
- Deduplicated content now in generic pages, resulting in cleanups in the TriplyDB.js and RATT documentation.
- Many small changes/additions for the TriplyDB.js documentation.
- Added most of the RATT documentation, based on sections that were previously published internally.
- I do not know how the changes to the 3 files related to blog posts got into this pull requests; these must have been introduced by others who committed before me in this feature branch.

rosalinedehaan committed:
added ratt directory

rosalinedehaan committed:
started on ducmentation

rosalinedehaan committed:
made note boldface

rosalinedehaan committed:
added assumptions for prior knowledge and some comments

rosalinedehaan committed:
added structure, text for 'Extract', 'Transform', 'Introdcutio'n and 'getting started' sections

rosalinedehaan committed:
published blog new features 2021-10

wouterbeek committed:
A big update for TriplyDB.js documentation.

1. The "Getting started" chapter is split into several sections.  They
start out with a read-only project and only set up a secure API token
later.  This keeps the number of steps small before the reader can see
something working.

2. The setting of a secure API token no longer assumes that the user
can install the direnv tool.  The documentation includes
installation/configuration steps that work on Windows, macOS and
Linux.

3. New sections on improved error handing and reporting.  These
explain the lengthy error-related code lines that appear in the start
script.

4. Included a Mermaid diagram that gives an overview of the main
classes and shows the relationships between them.

5. Added many sections on methods that were previously not documented.

6. Extended the description for almost all methods to include
information about access restrictions (with and without an API token),
examples, alternative approaches, "see also"-links to other parts of
the documentation.

7. The parts of the library that are unclear to me were specified with
"TODO" and were commented out with "<!-- … -->" notation.

8. All code snippets were adjusted to use the same example names and
the same notation.  Required awaits were added in many cases.

9. Documentation was removed for methods of the Organization and User
class that are inherited from the Account class.  The previous
documentation sections were merged into one, places under the Account
chapter, and linked to from the Organization and User chapters.

Commit:
Started with the ETL example for RATT, as a sort of hello-world example for RATT. I've tried to make it as small as possible but we have quite a bit of overhead when starting a RATT ETL.

Commit:
Added in example RATT ETL and made start in the core concepts.

wouterbeek committed:
ENHANCED: Processed the feedback by Rosaline.

Thanks for the good feedback!

wouterbeek committed:
ADDED: Rewrite of the “Getting Started” section in 3 subsections.

The 3 subsections describe increasingly more complex pipelines:

  1. Generate 1 statement to a local file.
  2. Publish 1 statement to a TriplyDB instance.
  3. Transform source data and upload to a TriplyDB instance.

wouterbeek committed:
ADDED: Generic pages about API Tokens and environment variables.

wouterbeek committed:
ADDED: A short introductory README file.

wouterbeek committed:
CLEAN: Removed duplicate descriptions of tokens and editors.

- ADDED: New page for how to configure text editors.

- CLEAN: Replace editor and token information from TriplyDB.js
         documentation with links to existing pages.

wouterbeek committed:
ADDED: Generic tips&tricks for writing clean code.

With examples from use in RATT.

wouterbeek committed:
ADDED: Section on upgrading RATT.

wouterbeek committed:
ADDED: Subsection on iterating over all assets as a RATT source.

wouterbeek committed:
FIXED: Used Google Docs for spellcheck.

wouterbeek committed:
FIXED: Used Google Docs for spellcheck.

wouterbeek committed:
CLEAN: Shorter identification for TypeScript code blocks.

wouterbeek committed:
ADDED: Stub for Client.isCompatibleWith.

wouterbeek committed:
ADDED: Section for Account.ensureDataset.

wouterbeek committed:
FIXED: Make JSON snippet valid JSON.

wouterbeek committed:
FIXED: Made running text consistent with code example for prefixes.

wouterbeek committed:
FIXED: Document that the role argument is optional.

wouterbeek committed:
CLEAN: Cleanup the beginning of the Service section.

wouterbeek committed:
ENHANCED: Completed the documentation for Dataset.importFromDataset.

- Added previously undocumented third argument.
- Included a comment for a possible future improvement.

wouterbeek committed:
FIXED: Consistenty call the functions on objects 'methods' in text.

wouterbeek committed:
CLEAN: Made it easier to distinguish subsections.

- 2 lines between method sections.

- 3 lines between object type / class sections.

- Also added some TODO items.

wouterbeek committed:
ADDED: Replaced a TODO item with an code example.

For `Dataset.getService`.

wouterbeek committed:
ADDED: Organization.ensureDataset and User.ensureDataset.

wouterbeek committed:
FIXED: Accounts cannot be deleted, only organization can be.

wouterbeek committed:
ADDED: Documented Dataset.getStatements.

wouterbeek committed:
CLEAN: Deleted outdated documentation for the Query object.

wouterbeek committed:
ADDED: Documented Story.delete.

wouterbeek committed:
TODO: Created stubs for the remaining undocumented methods (37).

wouterbeek committed:
ADDED: Integrated several RATT sections prepared internally.

These sections were previously added to the internal Triply
documentation and can now be added to the public documentation.

rosalinedehaan commented:
I took out the 3 files related to the blog post. I commited those to the wrong branch a while back.

MPvHarmelen committed:
Fix a/an in 4 places

evelienl commented on docs/ratt/index.md:
```diff
@@ -0,0 +1,1784 @@
+path: "/docs/ratt"
+---
+
+**RATT is distributed under the TriplyDB license.  Contact [info@triply.cc](mailto:info@triply.cc) for more information.**
```
and should not be used separately from a TriplyDB licence or TriplyDB SaaS product.

wouterbeek replied:
Adding this to the Management Team agenda.

evelienl commented on docs/ratt/index.md:
```diff
@@ -0,0 +1,1784 @@
+
+**RATT is distributed under the TriplyDB license.  Contact [info@triply.cc](mailto:info@triply.cc) for more information.**
+
+RATT is a [TypeScript package](https://www.npmjs.com/package/@triply/ratt) that is developed by Triply.  RATT makes it possible to develop and maintain production-grade linked data pipelines.  It is used in combination with a TriplyDB license and instance to create large-scale Knowledge Graphs.
```
knowledge graph without capitals

wouterbeek replied:
changed

rosalinedehaan committed:
added group, removed individual

MPvHarmelen commented on docs/ratt/index.md:
```diff
@@ -0,0 +1,1784 @@
+
+When we plot these two factors onto a table, we get the following overview of the RATT transformation functions:
+
+<figure>
+  <table>
+    <thead>
+      <tr><th></th><th>Create a new entry</th><th>Change an existing entry</th></tr>
+    </thead>
+    <tbody>
+      <tr><th>From the same entry</th><td>impossible</td><td><a href="#change"><code>change</code></a></td></tr>
+      <tr><th>From another entry</th><td><a href="#copy"><code>copy</code></a></td><td><a href="#replace"><code>replace</code></a></td></tr>
+      <tr><th>From the context</th><td><a href="#add"><code>add</code></a></td><td>Write a custom Middleware</td></tr>
+    </tbody>
+  </table>
+  <figcaption>Table - Overview of the RATT Transformation Functions.</figcaption>
+</figure>
```
Why is this an HTML table and not a markdown table?

See [here](https://github.com/TriplyDB/Documentation/blob/master/docs/triply-db/index.md?plain=1#L617) for the syntax. Maybe we should also add that we support markdown tables to the documentation that I linked.

wouterbeek replied:
Sorry, this is just too much work for me.  We need some documentation online.

MPvHarmelen commented on docs/ratt/index.md:
```diff
@@ -0,0 +1,1784 @@
+
+
+
+<h3 id="change">Change an existing entry in-place (<code>change</code>)</h3>
```
Same here, why is this HTML?

wouterbeek replied:
I have to refer to that section.  I do not know which ID is generated by Markdown.  I cannot test this locally.  Sorry.

MPvHarmelen commented on docs/ratt/index.md:
```diff
@@ -0,0 +1,1784 @@
+  mw.change({
+    key: 'KEY_NAME',
+    type: 'VALUE_TYPE',
+    change: value => FUNCTION_BODY}),
```
In the dev-team we consider it bad style not to have parentheses around the arguments of an arrow function, i.e. I would suggest `(value) => FUNCTION_BOYD`.

wouterbeek replied:
I'm looking into why brackets could be better and may change this later.

simonwtriply commented on docs/ratt/index.md:
```diff
@@ -0,0 +1,1784 @@
+
+Source data often contains 'special' values to denote the absence of a value.  Common examples are values such as `NULL` or the empty string (`""`) or 'clear' outliers such as `9999` for a missing year.  We call such values ‘null values’.
+
+The `mw.when` function supports the creation of triples under certain conditions.  The first argument that this function takes establishes whether or not a certain condition is met.  After that one or more additional statement arguments appear that will only be called if the condition is satisfied.  The generic structure of `mw.when` is as follows:
```
"After that, one or more..." (add comma)

wouterbeek replied:
Changed.

MPvHarmelen commented on docs/ratt/index.md:
```diff
@@ -0,0 +1,1784 @@
+  - `'literal'` an RDF literal term.
+  - `'number'` a natural number or floating-point number.
+  - `'string'` a sequence of characters.
+  - `'any'` an undetermined type.
```
Do we support that in RATT? And are we not documenting `unknown`?

wouterbeek replied:
I don't know what this means.  I have not seen `unknown` yet.

wouterbeek replied:
I hope somebody else will be able to fix this later.

MPvHarmelen commented on docs/ratt/index.md:
```diff
@@ -0,0 +1,1784 @@
+  - `'number'` a natural number or floating-point number.
+  - `'string'` a sequence of characters.
+  - `'any'` an undetermined type.
+- `FUNCTION_BODY` the body of a function, ending with a `return` statement that returns the new value.
```
This is not strictly speaking correct. In Typescript, you can have a shorthand arrow-function if the function would only have been a single return statement:

```ts
(ARGUMENTS) => EXPRESSION
```

is short for

```ts
(ARGUMENTS) => {
  return EXPRESSION
}
```

and arguably more readable.

wouterbeek replied:
I know very little about TS.  I have changed this into:
> - `FUNCTION_BODY` a function body that returns the new value.
But I have no idea whether this is more correct.

MPvHarmelen commented on docs/ratt/index.md:
```diff
@@ -0,0 +1,1784 @@
+mw.change({
+  key: 'Year',
+  type: 'string',
+  change: value => value.padStart(4, '0'),
```
We use the shorthand arrow-function notation here, so I do think we should document it (see comment above).

Again: `(value)` instead of `value`, to make it more explicit that we're introducing a new variable here, instead of using the value of an existing one?

wouterbeek replied:
I do not understand this comment.  Let's fix this later.

MPvHarmelen commented on docs/ratt/index.md:
```diff
@@ -0,0 +1,1784 @@
+mw.change({
+  key: 'Inhabitants',
+  type: 'unknown',
+  change: value => value as number / 1}),
```
Why not `+(value as number)`? Division may do weird stuff? Floats scare me.

wouterbeek replied:
I have never seen either `+` or `/` to cast a value.  Either way looks completely crazy to me, but happy to change to `+`.

MPvHarmelen commented on docs/ratt/index.md:
```diff
@@ -0,0 +1,1784 @@
+
+| Country     | Inhabitants |
+| ----------- | ----------- |
+| France      | null        |
```
If I try this in Node.js v14.18.1 the following happens:

```node
> '' / 1
0
> +''
0
```

Thus documenting the resulting value as `null` is incorrect.

wouterbeek replied:
Good catch!

wouterbeek replied:
Very strange: the empty strings is treated differently from other strings that do not encode a number.  I have adding this odd exception to the documentation and have also asked a question about this in the RATT chat channel.

evelienl commented on docs/ratt/index.md:
```diff
@@ -0,0 +1,1783 @@
+
+#### Standards-compliance
+
+RATT supports the official CSV standard: [RFC 4180](https://datatracker.ietf.org/doc/html/rfc4180).  Unfortunately, there are some ‘CSV’ files our there that do not follow the RFC 4180 standard.  Strictly speaking these files are not CSV files, although they may look similar to CSV files in some parts.
```
'our there' -> 'out there'

wouterbeek replied:
fixed

MPvHarmelen commented on docs/ratt/index.md:
````diff
@@ -0,0 +1,1784 @@
+
+```ts
+mw.change({
+  key: keys.country_of_birth,
````
`keys.country_of_birth` isn't explained anywhere. Should this be replaced by `'Source value'`?

wouterbeek replied:
Good catch!  changed.

TanjaRonzhina committed:
Update index.md

simonwtriply commented on docs/ratt/index.md:
```diff
@@ -0,0 +1,1784 @@
+- In case of an error, the first 20 values from the last processed RATT record.
+- In case of an error, the full stack trace.
+
+Example of using the `--verbose` flag:
```
"Example" doesn't start on a new line.

wouterbeek replied:
I do not understand what this means.  I have rewritten that sentence to:

> The following example shows how the `--verbose` flag can be used:

I hope this fixes the issue that you saw.

TanjaRonzhina committed:
Update index.md

Added lines what lines of the code do inside the code block

MPvHarmelen committed:
Small edits to RATT

MPvHarmelen committed:
Hopefully fix code highlighting

MPvHarmelen reviewed (changes requested):
See also my changes in c1ff06bc67c4458d6bbfa00d5052cbd570f6a871

TanjaRonzhina committed:
Update index.md

Added lines describing the code into the code block 
corrected spelling

TanjaRonzhina committed:
Update index.md

Corrected spelling and typos at RATT external documentation

TanjaRonzhina committed:
Update index.md

Corrected typos in ratt documentation

TanjaRonzhina committed:
Update index.md

Corrected spelling and typos at ratt external documentation

TanjaRonzhina committed:
Update index.md

Corrected links, typos, style at ratt external documentation

TanjaRonzhina committed:
Update index.md

Small styling changes at external ratt documentation

TanjaRonzhina committed:
Update index.md

Corrected link

TanjaRonzhina committed:
Update index.md

Updated the link

wouterbeek committed:
FIXED: Processed feedback on my pull request.  Thank you to the reviewers!

wouterbeek reviewed (commented):
LGTM

wouterbeek reviewed (commented):
LGTM

wouterbeek commented:
I cannot finish this pull request because I lack knowledge of Git/Github.  I hope somebody else can take over of do this together with me.

TanjaRonzhina commented:
Double checked the documentation. 
LGTM!
Only the broken links should be updated:
1. in section [External prefixes](https://github.com/TriplyDB/Documentation/blob/feature/ratt/docs/ratt/index.md#external-prefixes),  [the link](https://github.com/TriplyDB/Documentation/blob/feature/ratt/docs/generics/editor) in the sentence "In properly configured text editors this will bring up a list of autocomplete results" doesn't work
2. in section [Publish to TriplyDB](https://github.com/TriplyDB/Documentation/blob/feature/ratt/docs/ratt/index.md#publish-to-triplydb) link [how to configure API token](https://github.com/TriplyDB/Documentation/blob/feature/ratt/docs/generics/api-token) doesn't work
3. in section [verbose mode](https://github.com/TriplyDB/Documentation/blob/feature/ratt/docs/ratt/index.md#verbose-mode) link to [environment variables](https://github.com/TriplyDB/Documentation/blob/feature/ratt/docs/generics/environment-variable) also ends with 404 error

wouterbeek committed:
Merge branch 'master' into feature/ratt

wouterbeek committed:
FIXED: Many small fixes and cleanups to RATT documentation.

Based on:
- Feedback from many Triplies.
- Live local website fixed for me by Gerwin.
- Google Docs copy/paste spellcheck (alternatively use a spellcheck
  plugin in Atom).
- Missing links noticed by Tanja.
- Commented out all TODO items.
